File: src/main/java/org/hisp/dhis/rules/gs1/GS1DataMatrixValueFormatter.java
package org.hisp.dhis.rules.gs1;

import java.util.HashMap;
import java.util.Map;

public class GS1DataMatrixValueFormatter extends GS1BaseValueFormatter
{
    private Map<String, String> dataMap;

    public String formatValue( String value, GS1Elements valueToReturn )
    {
        dataMap = new HashMap<>();
        String[] gs1Groups = removeGS1Identifier( value ).split( GS1Elements.GS1_GROUP_SEPARATOR.getElement() );
        for ( String gs1Group : gs1Groups )
        {
            handleGroupData( gs1Group );
        }
        if ( dataMap.containsKey( valueToReturn.getElement() ) )
        {
            return dataMap.get( valueToReturn.getElement() );
        }
        else
        {
            throw new IllegalArgumentException( "Required key does not exist for provided value" );
        }
    }

    private void handleGroupData( String gs1Group )
    {
        if ( !gs1Group.isEmpty() )
        {
            int gs1GroupLength = gs1Group.length();
            String ai = GS1Elements.getApplicationIdentifier( gs1Group );
            Integer nextValueLength = GS1Table.aiFixedLengthMap().get( ai );
            if ( nextValueLength == null )
                nextValueLength = gs1GroupLength;
            dataMap.put( ai, gs1Group.substring( 2, nextValueLength ) );
            handleGroupData( gs1Group.substring( nextValueLength ) );
        }
    }
}

File: src/commonMain/kotlin/org.dhis2.ruleengine/gs1/GS1DataMatrixValueFormatter.kt
package org.dhis2.ruleengine.gs1

import org.dhis2.ruleengine.gs1.GS1Table.aiFixedLengthMap

class GS1DataMatrixValueFormatter : GS1BaseValueFormatter() {
    private var dataMap: MutableMap<String, String>? = null
    override fun formatValue(value: String?, valueToReturn: GS1Elements): String? {
        dataMap = mutableMapOf()
        val gs1Groups = removeGS1Identifier(value)!!.split(GS1Elements.GS1_GROUP_SEPARATOR.element.toRegex())
            .dropLastWhile { it.isEmpty() }
            .toTypedArray()
        for (gs1Group in gs1Groups) {
            handleGroupData(gs1Group)
        }
        return if (dataMap!!.containsKey(valueToReturn.element)) {
            dataMap!![valueToReturn.element]
        } else {
            throw IllegalArgumentException("Required key does not exist for provided value")
        }
    }

    private fun handleGroupData(gs1Group: String) {
        if (!gs1Group.isEmpty()) {
            val gs1GroupLength = gs1Group.length
            val ai = GS1Elements.getApplicationIdentifier(gs1Group)
            var nextValueLength = aiFixedLengthMap()[ai.substring(0, 2)]
            if (nextValueLength == null) nextValueLength = gs1GroupLength
            dataMap!![ai] = gs1Group.substring(ai.length, nextValueLength)
            handleGroupData(gs1Group.substring(nextValueLength))
        }
    }
}